Add `torch._C._set_watchdog()` that spawns a pthread-watchdog

malfet opened the pull request:
Hopefully that should be reliable enough to detect any native/managed deadlocks while running python testsuite

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
+     watchdogs.erase(it);
+     Py_RETURN_NONE;
+   }
+   std::cerr << " Ignoring watchdog override " << name << std::endl;
```
lol

ezyang replied:
Can we just make this a hard error?

malfet replied:
Don't have strong preference, one way or the other, will add `std::abort()` and see what will happen.

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
+      cv.wait_for(lock, timeout *1s, [this] { return exitFlag; });
+      if (!exitFlag) {
+        std::cerr << "Test " << name << " takes longer than " << timeout << " seconds" <<std::endl;
+        std::exit(-1);
```
Are we sure we want a `std::exit` here? Maybe a `std::abort` would be better at actually killing the process?

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
+  WatchdogThread(const std::string& name_, int timeout): name(name_), thrd([this, timeout] {
+      using namespace std::chrono_literals;
+      std::unique_lock<std::mutex> lock(mutex);
+      cv.wait_for(lock, timeout *1s, [this] { return exitFlag; });
```
Second suffix, so fancy!!

malfet replied:
I like `chrono_literls` :)

ezyang commented on torch/csrc/Module.cpp:
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
+  std::mutex mutex;
+  std::condition_variable cv;
+  bool exitFlag = false;
+  // Thread must be the last member, otherwise it will not get properly initialized
```
because you need the cv/mutex etc to be initialized before it, is that right? Subtle...

malfet replied:
Yes, both `mutex`, `cv` and `exitFlag` must be initialized before thread is created. (I knew about this, but actually it was a bug in my initial implementation and this race was hit half way thru `torch_jit` tests)

malfet committed:
Add `torch._C._set_watchdog()` that spawns a pthread-watchdog

Hopefully that should be reliable enough to detect any native/managed deadlocks while running python testsuite
